libs/wingc/src/jsify.rs

```diff
@@ -1771,87 +1764,115 @@ impl<'a> FieldReferenceVisitor<'a> {
 	}
 }
 
-/// Visitor that finds all the types defined in preflight that
-/// are referenced inside a scope (used on inflight methods)
-struct PreflightTypeRefVisitor<'a> {
+/// Visitor that finds all the types and variables used within an inflight method but defined in its
+/// parent environment
+struct CaptureScanner<'a> {
 	/// Set of user types referenced inside the method
-	references: IndexMap<String, TypeRef>,
+	captured_types: IndexMap<String, TypeRef>,
+
+	/// The set of free variables referenced by the method
+	captured_vars: IndexSet<String>,
 
 	/// The root scope of the function we're analyzing
 	function_scope: &'a Scope,
 
-	/// The current env, used to lookup the type
-	env: SymbolEnvRef,
+	/// The current method env, used to lookup the type
+	method_env: SymbolEnvRef,
+
+	/// Tracks the statement index
+	statement_index: usize,
 }
 
-impl<'a> PreflightTypeRefVisitor<'a> {
+impl<'a> CaptureScanner<'a> {
 	pub fn new(function_scope: &'a Scope) -> Self {
 		Self {
-			references: IndexMap::new(),
+			captured_types: IndexMap::new(),
+			captured_vars: IndexSet::new(),
 			function_scope,
-			env: function_scope.env.borrow().as_ref().unwrap().get_ref(),
+			method_env: function_scope.env.borrow().as_ref().unwrap().get_ref(),
+			statement_index: 0,
 		}
 	}
 
-	pub fn find_preflight_type_refs(mut self) -> IndexMap<String, TypeRef> {
+	pub fn scan(&mut self) {
 		self.visit_scope(self.function_scope);
-		self.references
 	}
 
-	fn add_ref(&mut self, type_name: String) {
-		// Lookup the type in the current env and see where it was defined
-		if let LookupResult::Found(kind, _) = (*self.env).lookup_nested_str(&type_name, None) {
-			// This must be a type reference
-			if let SymbolKind::Type(t) = kind {
-				// If this class was defined preflight then store it
-				let decl_phase = t.as_class().and_then(|f| Some(f.declaration_phase)).unwrap_or(Phase::Preflight);
-				if decl_phase == Phase::Preflight {
-					self.references.insert(type_name, *t);
+	fn consider_reference(&mut self, path: &[&Symbol]) {
+		if let LookupResult::Found(kind, info) = self.method_env.lookup_nested(path, Some(self.statement_index), 0) {
+			// if the symbol was found in our current enviroment, we move on. that's the essence of capturing:
+			// you reference something that defined in a parent environment.
+			if info.depth == 0 {
```

Comment:
As we discussed this needs to account for the case of a nested scope inside the function.

Reply:
done